### PR TITLE
Vmware guest enhancements

### PR DESCRIPTION
**Pull request by aperigault**
##### ISSUE TYPE
 - Feature Pull Request
 - Bugfix Pull Request

##### SUMMARY
I added vlan support and correct minor bugs.
I think you need to rebase your branch to see a better diff.

### CODE REVIEW

**Comment by nerzhul** [comment]
i rebased the branch, but did a commit this morning without seeing your PR, can you rebase again ?

Sorry for the inconvenience

**Comment by aperigault** [comment]
@nerzhul Done :-)

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 1]
```diff
@@ -833,18 +833,30 @@ def configure_network(self, vm_obj):
...
+                    network_name = None
+                    dvps = get_all_objs(self.content, [vim.dvs.DistributedVirtualPortgroup])
+                    for dvp in dvps:
+                        if dvp.config.defaultPortConfig.vlan.vlanId == self.params['networks'][network]['vlan']:
+                            network_name = dvp.config.name
+                            break
+                    if network_name:
+                        self.params['networks'][network]['network'] = network_name
+                    else:
+                        self.module.fail_json(msg="vlan %s doesn't exists" % self.params['networks'][network]['vlan'])
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py]: Can you make vlan upper case for labels ?

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 2]
```diff
@@ -1103,7 +1116,7 @@ def select_datastore(self, vm_obj=None):
 
     def obj_has_parent(self, obj, parent):
         assert obj is not None and parent is not None
-        current_parent = obj.parent
+        current_parent = obj
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1119]: nice

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 3]
```diff
@@ -1163,6 +1176,8 @@ def deploy_vm(self):
 
         # find matching folders
         if self.params['folder'].startswith('/'):
+            if not self.params['folder'].startswith('/vm'):
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:1179]: What does it fixes ?

File: lib/ansible/modules/cloud/vmware/vmware_guest.py [hunk 4]
```diff
@@ -908,19 +920,20 @@ def configure_network(self, vm_obj):
                 self.change_detected = True
 
             if vm_obj is None or self.should_deploy_from_template():
-                guest_map = vim.vm.customization.AdapterMapping()
-                guest_map.adapter = vim.vm.customization.IPSettings()
-                guest_map.adapter.ip = vim.vm.customization.FixedIp()
-                guest_map.adapter.ip.ipAddress = str(network_devices[key]['ip'])
-                guest_map.adapter.subnetMask = str(network_devices[key]['subnet_mask'])
+                if 'ip' in self.params['networks'][network]:
```
Comment (nerzhul) [inline lib/ansible/modules/cloud/vmware/vmware_guest.py:923]: With this you make ip optionnal, i didn't test the case, it's working without this field when creating a VM from scratch ?

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 3 above)
Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py:1179]: There is a similar thing in getvm function. /vm/ is a masked folder so we add it in path if it's not set.
It's allows to define full path without vmware knowledges.

Comment (nerzhul) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py:1179]: okay, i'm just waiting for answer on other points and the uppercase issue and then i will merge it

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 4 above)
Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py:923]: I just tested this use case now and it seems to be ok. I can create a VM from scratch with this network parameters:
```
    networks:
      my_network_interface:
        vlan: 163
```

File: lib/ansible/modules/cloud/vmware/vmware_guest.py (hunk 1 above)
Comment (aperigault) [reply lib/ansible/modules/cloud/vmware/vmware_guest.py]: Fixed

**Comment by nerzhul** [comment]
merged :)

**Pull request merged**